src/interpreter/processor.rb

```diff
@@ -0,0 +1,212 @@
+require_relative '../colour_string'
+require_relative '../token'
+require_relative '../util/logger'
+require_relative '../util/options'
+
+require_relative 'errors'
+require_relative 'formatter'
+require_relative 'return_value'
+require_relative 'scope'
+
+require_relative 'processor/built_ins'
+Dir["#{__dir__}/processor/token_processors/*.rb"].each { |f| require_relative f }
+
+module Interpreter
+  class Processor
+    MAX_CALL_STACK_DEPTH = 1000
+
+    ############################################################################
+    # TokenProcessors consist of processing methods and any processing-specific
+    # helper methods.
+    #
+    # Processing:
+    # These methods take tokens and execute them. They may also read from the
+    # stack or additional tokens in the token stream. The stack should be clear
+    # after each call to a processing method.
+    #
+    # Any other common or helper methods will be in this file.
+    ############################################################################
+    include TokenProcessors
+
+    include BuiltIns
+
+    def initialize(lexer, options = {})
+      @lexer   = lexer
+      @options = options
+
+      @current_scope = Scope.new
+
+      # The current stack of tokens which have not been processed.
+      @stack = []
+    end
+
+    def execute
+      process
+    rescue Errors::BaseError => e
+      e.line_num = @lexer.line_num
+      raise
+    end
+
+    private
+
+    def process
+      raise Errors::CallStackTooDeep, MAX_CALL_STACK_DEPTH if caller.length > MAX_CALL_STACK_DEPTH
+
+      loop do
+        token = next_token
+        break if token.nil?
+
+        result = process_token token
+        return result if result.is_a? ReturnValue
+      end
+
+      nil
+    end
+
+    def next_token
+      token = peek_next_token
+      Util::Logger.debug Util::Options::DEBUG_2, 'RECEIVE: '.lred + (token ? "#{token} #{token.content}" : 'EOF')
+      @current_scope.advance
+      token
+    end
+
+    def peek_next_token
```

Comment:
Doc comment:

> If this is the main scope, tokens are simply read from the lexer and discarded. Otherwise, the tokens of other types of scopes are stored in their bodies and kept track of using a token pointer.
> This function simply returns the current token under the token pointer, without advancing it.